### ISSUE TITLE
Can't build BOLT on a aarch64 machine because runtime/common.h uses x86_64-specific code

### ISSUE THREAD

**Issue by Romain-Geissler-1A** [facebookincubator/BOLT#102]
Hi,

I am trying to migrate an existing toolchain (containing BOLT) from a x86_64 host/target to a aarch64 host/target one. I was hoping Bolt was buildable on a aarch64 system given that it can handle aarch64 binaries. But apparently it's not the case, as at least runtime/common.h seems to contain code referring to registers existing only on x86_64 machines. I have the following compilation error:

```
/workdir/src/llvm-old-repo-7.0.0/tools/llvm-bolt/runtime/common.h: In function 'bool has_pagecache_thp_support()':
/workdir/src/llvm-old-repo-7.0.0/tools/llvm-bolt/runtime/common.h:127:55: error: unknown register name 'rcx' in 'asm'
                        : "cc", "rcx", "r11", "memory");
                                                       ^
/workdir/src/llvm-old-repo-7.0.0/tools/llvm-bolt/runtime/common.h: In function 'void __bolt_hugify_self_impl()':
/workdir/src/llvm-old-repo-7.0.0/tools/llvm-bolt/runtime/common.h:167:55: error: unknown register name 'rcx' in 'asm'
                        : "cc", "rcx", "r11", "memory");
                                                       ^
/workdir/src/llvm-old-repo-7.0.0/tools/llvm-bolt/runtime/common.h:167:55: warning: asm operand 2 probably doesn't match constraints
/workdir/src/llvm-old-repo-7.0.0/tools/llvm-bolt/runtime/common.h:74:55: error: unknown register name 'rcx' in 'asm'
                        : "cc", "rcx", "r11", "memory");
                                                       ^
/workdir/src/llvm-old-repo-7.0.0/tools/llvm-bolt/runtime/common.h:74:55: warning: asm operand 2 probably doesn't match constraints
/workdir/src/llvm-old-repo-7.0.0/tools/llvm-bolt/runtime/common.h:217:55: error: unknown register name 'rcx' in 'asm'
                        : "cc", "rcx", "r11", "memory");
                                                       ^
/workdir/src/llvm-old-repo-7.0.0/tools/llvm-bolt/runtime/common.h:167:55: error: unknown register name 'rcx' in 'asm'
                        : "cc", "rcx", "r11", "memory");
                                                       ^
```

It is expected that we can't build Bolt itself on aarch64 ?

Cheers,
Romain

**Comment by maksfb** [comment]
Hi Romain, It is somewhat expected as the runtime library is currently supported on x86-64 only. The fix will be to exclude the library from the build on a different architecture or use a cross-compiler with x86-64 target support. If aarch64 is your primary target, and you plan to use instrumentation, then you'll also need to create aarch64 version of the instrumentation library.

**Comment by Romain-Geissler-1A** [comment]
Hi,

Thanks for the reply. I really tro try to compile Bolt on a aarch64 to be able to run it on a aarch64 machine and optimize aarch64 binaries. I naively thought it would be ok, but ok, apparently for now Facebook's seems to do these instrumentation operations just on x86-64. I guess Facebook has no plan to migrate their servers to aarch64 yet ;)

Since right now this is expected, and since I guess you don't really have plan to ever make it work without x86-64, I think we should just close this ticket ?

Cheers,
Romain

**Comment by maksfb** [comment]
TL;DR: if you can run `perf` with `-e cycles` you should be able to run BOLT on aarch64.

You don't need the instrumentation if you have access to hardware counters via perf. Preferably, these should be last branch records or anything similar, but you can also use simple events like cycles. When using cycles, you have to add `-nl` flag to `perf2bolt`.

**Comment by Romain-Geissler-1A** [comment]
Ok thanks

**Comment by ijinma** [comment]
> 你好，
> 
> 谢谢回复。我真的尝试在aarch64上编译Bolt，以便能够在aarch64机器上运行它并优化aarch64二进制文件。我天真地以为可以，但是现在看来，Facebook似乎只是在x86-64上执行这些检测操作。我猜Facebook还没有计划将其服务器迁移到aarch64；）
> 
> 既然现在是可以预料的，并且由于我猜您真的没有计划要在没有x86-64的情况下工作，所以我认为我们应该关闭这张票吗？
> 
> 欢呼声，
> 罗曼



> Hi,
> 
> Thanks for the reply. I really tro try to compile Bolt on a aarch64 to be able to run it on a aarch64 machine and optimize aarch64 binaries. I naively thought it would be ok, but ok, apparently for now Facebook's seems to do these instrumentation operations just on x86-64. I guess Facebook has no plan to migrate their servers to aarch64 yet ;)
> 
> Since right now this is expected, and since I guess you don't really have plan to ever make it work without x86-64, I think we should just close this ticket ?
> 
> Cheers,
> Romain

Hi，

  I have a similar situation, have you solved it and run it successfully on aarch64? What should I do?

  thanks，
  ijinma

**Comment by ijinma** [comment]
> Hi,
> 
> Thanks for the reply. I really tro try to compile Bolt on a aarch64 to be able to run it on a aarch64 machine and optimize aarch64 binaries. I naively thought it would be ok, but ok, apparently for now Facebook's seems to do these instrumentation operations just on x86-64. I guess Facebook has no plan to migrate their servers to aarch64 yet ;)
> 
> Since right now this is expected, and since I guess you don't really have plan to ever make it work without x86-64, I think we should just close this ticket ?
> 
> Cheers,
> Romain

Hi，

I have a similar situation, have you solved it and run it successfully on aarch64? What should I do?

thanks，
ijinma

**Comment by Romain-Geissler-1A** [comment]
No on my side I just gave up trying on aarch64 hosts.

**Comment by ijinma** [comment]
> Hi Romain, It is somewhat expected as the runtime library is currently supported on x86-64 only. The fix will be to exclude the library from the build on a different architecture or use a cross-compiler with x86-64 target support. If aarch64 is your primary target, and you plan to use instrumentation, then you'll also need to create aarch64 version of the instrumentation library.

hi, @maksfb 
  I removed the runtime that generated the error message (```runtime/common.h:127:55: error: unknown register name 'rcx' in 'asm' : "cc", "rcx", "r11", "memory" ```)from the llvm-bolt so that the bolt was successfully compiled and run on arrch64 host. 
  After testing, the bolt can optimize mysql. I wonder if I'm going to have any side effects on bolt in this way.

thanks，
ijinma
